cl/assets/static-global/js/progress-bar.js

```diff
@@ -0,0 +1,47 @@
+let trickleInterval = null;
+
+function updateProgressBar() {
+  let progressElement = document.getElementById('progress-bar');
+  let oldValue = 0;
+  if ('value' in progressElement.dataset) {
+    oldValue = parseFloat(progressElement.dataset.value);
+  }
+  let newValue = oldValue + Math.random() / 10;
+  progressElement.style.width = `${10 + newValue * 85}%`;
+  progressElement.dataset.value = newValue;
+}
+
+function installProgressBar() {
+  let progressElement = document.createElement('div');
+  progressElement.id = 'progress-bar';
+  progressElement.classList.add('turbo-progress-bar');
+  progressElement.style.width = '0';
+  progressElement.style.opacity = '1';
+  document.body.prepend(progressElement);
+  trickleInterval = window.setInterval(updateProgressBar, 300);
+}
+
+function disableAllSubmitButtons() {
+  // Get all submit buttons on the page
+  const submitButtons = document.querySelectorAll('input[type="submit"],button[type="submit"]');
+  // Disable each element
+  submitButtons.forEach((button) => {
+    button.disabled = true;
+  });
+}
+
+document.onvisibilitychange = () => {
```

Comment:
Can you help me understand why you're doing it this way?

Reply:
Without `HTMX` events to track request completion, we needed a way to update the progress bar(hide the bar) as the browser receives the server response and begins loading the next page. My initial exploration led me to the `beforeunload` event, which fires right before the user leaves the page. While this event seemed promising, the [MDN Web Docs](https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event#usage_notes) revealed its unreliability, particularly on mobile platforms. They recommend the [visibilitychange event](https://developer.mozilla.org/en-US/docs/Web/API/Document/visibilitychange_event) as a more dependable signal.

Reply:
Interesting. I would have expected it to not need anything? Like, the whole page is loading, so when the new one comes, the bar gets nuked. I guess this is slightly faster because we get the signal a bit sooner? 

I was also curious how this is impacted by people like me that click between tabs a lot while waiting for things to load, but I wasn't able to get a slow query to see if there was an impact of some kind.